smithy-jsonschema/src/main/java/software/amazon/smithy/jsonschema/JsonSchemaConverter.java

```diff
@@ -109,216 +151,270 @@ public ObjectNode getConfig() {
-
+    public SchemaDocument convert() {
+        LOGGER.fine("Converting to JSON schema");
         SchemaDocument.Builder builder = SchemaDocument.builder();
-        JsonSchemaShapeVisitor visitor = new JsonSchemaShapeVisitor(
-                model, getConfig(), getRefStrategy(), getPropertyNamingStrategy(), mappers);
 
         if (rootShape != null && !(rootShape instanceof ServiceShape)) {
+            LOGGER.fine(() -> "Setting root schema to " + rootShape);
             builder.rootSchema(rootShape.accept(visitor));
         }
 
         addExtensions(builder);
-        Predicate<Shape> predicate = composePredicate(model, rootShape);
+
+        // Create a model that strips out traits and disconnected shapes.
+        Model updatedModel = createUpdatedModel(model, rootShape, shapePredicate);
+
         model.shapes()
-                .filter(predicate)
-                // Don't include members if their container was excluded.
-                .filter(shape -> memberDefinitionPredicate(model, shape, predicate))
+                // Only generate shapes that passed through each predicate.
+                .filter(shape -> updatedModel.getShape(shape.getId()).isPresent())
+                // Don't generate members.
+                .filter(FunctionalUtils.not(Shape::isMemberShape))
+                // Don't write the root shape to the definitions.
+                .filter((shape -> rootShape == null || !shape.getId().equals(rootShape.getId())))
+                // Don't convert unsupported shapes.
+                .filter(FunctionalUtils.not(this::isUnsupportedShapeType))
+                // Ignore prelude shapes.
+                .filter(FunctionalUtils.not(Prelude::isPreludeShape))
+                // Do not generate inlined shapes in the definitions map.
+                .filter(FunctionalUtils.not(refStrategy::isInlined))
+                // Create a pair of pointer and shape.
+                .map(shape -> Pair.of(toPointer(shape), shape))
+                // Only add definitions if they are at the top-level and not inlined.
+                .filter(pair -> isTopLevelPointer(pair.getLeft()))
                 // Create the pointer to the shape and schema object.
-                .map(shape -> Pair.of(
-                        getRefStrategy().toPointer(shape.getId(), getConfig()),
-                        shape.accept(visitor)))
+                .map(pair -> {
+                    LOGGER.fine(() -> "Converting " + pair.getRight() + " to JSON schema at " + pair.getLeft());
+                    return Pair.of(pair.getLeft(), pair.getRight().accept(visitor));
+                })
                 .forEach(pair -> builder.putDefinition(pair.getLeft(), pair.getRight()));
 
+        LOGGER.fine(() -> "Completed JSON schema document conversion (root shape: " + rootShape + ")");
+
+        // Add extensions if present.
+        config.getObjectMember(JsonSchemaConstants.SCHEMA_DOCUMENT_EXTENSIONS).ifPresent(builder::extensions);
```

Comment:
This is doing the [same work as this line](https://github.com/awslabs/smithy/pull/274/files#diff-aaa37e69f32f0a6ca741784407a1fc55R220), just without going to the function.